Add WebCodecs article. (#4022)
GoogleChrome/web.dev: src/site/content/en/blog/web-codecs/index.md

---
title: Video processing with WebCodecs
subhead: Manipulating video stream components
description: |
  Work with components of a video stream, such as frames and unmuxed chunks of encoded video or audio.
date: 2020-10-13
updated: 2020-10-13
authors:
 - djuffin
origin_trial:
    url: https://developers.chrome.com/origintrials/#/view_trial/-7811493553674125311
tags:
  - blog
  - media
  - video
feedback:
  - api
---

Modern web technologies provide ample ways to work with video.
[Media Stream API](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream_Recording_API),
[Media Recording API](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream_Recording_API)
,
[Media Source API](https://developer.mozilla.org/en-US/docs/Web/API/Media_Source_Extensions_API),

> [WebRTC API](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API) add up

Reviewer: ```suggestion
 and [WebRTC API](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API) add up
```

screen.

```js
let frames_from_stream = [];
let stream = await navigator.mediaDevices.getUserMedia({ ... });
let vtr = new VideoTrackReader(stream.getVideoTracks()[0]);
vtr.start((frame) => {
  frames_from_stream.push(frame);
});
```

No matter where they are coming from, frames can be encoded into
`EncodedVideoChunk` objects with a `VideoEncoder`.

Before encoding `VideoEncoder` needs to be given two javascript objects:

+   Init dictionary with two functions for handling encoded chunks and
    errors. These functions are developer-defined and can't be changed after
    they're passed to the `VideoEncoder` constructor..
+   Encoder configuration object, which contains parameters for the output
    video stream. You can change these parameters later  by calling `configure()`.

```js
const init = {
  output: handleChunk,
  error: (e) => {
    console.log(e.message);
  }
};

let config = {
  codec: "vp8",
  width: 640,
  height: 480,
  bitrate: 8_000_000,     // 8 Mbps
  framerate: 30,
};

let encoder = new VideoEncoder(init);
encoder.configure(config);
```

After the encoder has been set up,  it's ready to start accepting frames. When

> frames are coming from a media stream, `VideoTrackReader` will pump frames into

Reviewer: ```suggestion
frames are coming from a media stream, the `VideoTrackReader` callback will pump frames into
```

Author: What is he trying to communicate? `VideoTrackReader` is an interface, not a callback.

Author: Is he trying to refer to the callback passed to `VideoTrackReader.start()`? If so, it needs to be that explicit. What he suggested is too big of a leap for people don't know anything about WebCodecs (which is anyone reading this article).

Reviewer: Yeah, that's exactly what he means here.  How about this  
```suggestion
frames are coming from a media stream, the callback given to `VideoTrackReader.start()` will pump frames into
```

Feel free to reword it in whatever way you see fit.

vtr.start((frame) => {
  if (pending_outputs > 30) {
    // Too many frames in flight, encoder is overwhelmed
    // let's drop this frame.
    return;
  }
  frame_counter++;
  pending_outputs++;
  const insert_keyframe = (frame_counter % 150) == 0;
  encoder.encode(frame, { keyFrame: insert_keyframe });
});
```

Finally it's time to finish encoding code by writing a function that handles
chunks of encoded video as they come out of the encoder.
Usually this function would be sending data chunks over the network or [muxing
](https://en.wikipedia.org/wiki/Multiplexing#Video_processing)them into a media
container for storage.

```js
function handleChunk(chunk) {
  let data = new Uint8Array(chunk.data);  // actual bytes of encoded data
  let timestamp = chunk.timestamp;        // media time in microseconds
  let is_key = chunk.type == "key";       // can also be "delta"
  pending_outputs--;
  fetch('/upload_chunk?timestamp=' + timestamp + "&type=" + chunk.type,
  {
    method: 'POST',
    headers: {   'Content-Type': 'application/octet-stream'   },
    body: data
  });
}
```

If at some point you'd need to make sure that all pending encoding requests have

> been completed, you can call `flush()` and wait for its promise. For example,

Reviewer: ```suggestion
been completed, you can call `flush()` and wait for its promise.
```

> it's required to call `flush()` before the encoder can be reconfigured.

Reviewer: ```suggestion
```

  return Math.max(0, (timestamp / 1000) - media_time);
}

async function render_frame() {
  if (ready_frames.length == 0) {
    underflow = true;
    return;
  }
  let frame = ready_frames.shift();
  underflow = false;

  let bitmap = await frame.createImageBitmap();
  frame.destroy();
  // Based on the frame's timestamp calculate how much of real time waiting
  // is needed before showing the next frame.
  let time_till_next_frame = calculateTimeTillNextFrame(frame.timestamp);
  await delay(time_till_next_frame);
  ctx.drawImage(bitmap, 0, 0);

  // Immediately schedule rendering of the next frame
  setTimeout(render_frame, 0);
}
```

## Demo

The demo below shows two canvases, the first one is animated at the refresh rate
of your display, the second one shows a sequence of frames captured by
`VideoTrackReader` at 30fps, encoded  and decoded using WebCodecs API.

[https://webcodecs-blogpost-demo.glitch.me/](https://webcodecs-blogpost-demo.glitch.me/)

## Feature Detection

This is an easy way to check for WebCodecs support, to make  sure that the
origin trial or the flag took effect.

```js
if ("VideoEncoder" in window) {
  // WebCodecs API is supported.
}
```

> ## Using the WebHID API {: #use }

Reviewer: ```suggestion
## Using the WebCodecs API {: #use }
```

> ### Enabling via chrome://flags

Reviewer: There is no way to enable WebCodecs via chrome://flags. 
Please preserve original wording:

```
To enable WebCodecs locally, start Chrome with a command line flag: --enable-blink-features=WebCodecs.
```

Author: Fixed.

> To experiment with the WebHID API locally on all desktop platforms, without an

Reviewer: ```suggestion
To experiment with the WebCodecs API locally on all desktop platforms, without an
```

origin trial token, start Chrome with a command line flag:

```bash
--enable-blink-features=WebCodecs.
```

### Enabling support during the origin trial phase

The WebCodecs API is available on all desktop platforms (Chrome OS, Linux, macOS,
and Windows) as an origin trial in Chrome&nbsp;86. The origin trial is expected
to end just before Chrome&nbsp;88 moves to stable in February 2021. The API can
also be enabled using a flag.

> {% include 'content/origin-trials.njk' %}

Reviewer: Does this template provide a link to the right origin trial?

Author: Yes. It reads the URL from the metadata at the top.

  let bitmap = await frame.createImageBitmap();
  frame.destroy();
  // Based on the frame's timestamp calculate how much of real time waiting
  // is needed before showing the next frame.
  let time_till_next_frame = calculateTimeTillNextFrame(frame.timestamp);
  await delay(time_till_next_frame);
  ctx.drawImage(bitmap, 0, 0);

  // Immediately schedule rendering of the next frame
  setTimeout(render_frame, 0);
}
```

## Demo

The demo below shows two canvases, the first one is animated at the refresh rate
of your display, the second one shows a sequence of frames captured by
`VideoTrackReader` at 30fps, encoded  and decoded using WebCodecs API.

[https://webcodecs-blogpost-demo.glitch.me/](https://webcodecs-blogpost-demo.glitch.me/)

## Feature Detection

This is an easy way to check for WebCodecs support, to make  sure that the
origin trial or the flag took effect.

```js
if ("VideoEncoder" in window) {
  // WebCodecs API is supported.
}
```

## Using the WebCodecs API {: #use }

### Enabling via a command line flag

To experiment with the WebCodecs API locally on all desktop platforms, without an
origin trial token, start Chrome with a command line flag:

```bash

> --enable-blink-features=WebCodecs.

Reviewer 2: ```suggestion
--enable-blink-features=WebCodecs

---
title: Video processing with WebCodecs
subhead: Manipulating video stream components.
description: |
  Work with components of a video stream, such as frames and unmuxed chunks of encoded video or audio.
date: 2020-10-13

> updated: 2020-10-13

Reviewer 3: ```suggestion
# updated: 2020-10-13
```
I usually omit this until we actually update the page

hero: hero.jpg
alt: A roll of film.
authors:
 - djuffin
origin_trial:
    url: https://developers.chrome.com/origintrials/#/view_trial/-7811493553674125311
tags:
  - blog
  - media
  - video
feedback:
  - api
---

Modern web technologies provide ample ways to work with video.
[Media Stream API](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream_Recording_API),

> [Media Recording API](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream_Recording_API)
> ,

Reviewer 3: ```suggestion
[Media Recording API](https://developer.mozilla.org/en-US/docs/Web/API/MediaStream_Recording_API),
```
Nitpick

[`ImageBitmap`](https://developer.mozilla.org/en-US/docs/Web/API/ImageBitmap).
Just call the `VideoFrame()` constructor and give it a bitmap and a presentation
timestamp.

```js
let cnv = document.createElement('canvas');
// draw something on the canvas
…
let bitmap = await createImageBitmap(cnv);
let frame_from_bitmap = new VideoFrame(bitmap, { timestamp: 0 });
```

<figure class="w-figure">
  <img src="./consuming-frames-from-bitmaps.png" alt="The path from ImageBitmap to the network or to storage.">
  <figcaption class="w-figcaption">The path from <code>ImageBitmap</code> to the network or to storage.</figcaption>
</figure>

The second is to use `VideoTrackReader` to set a function that will be called
each time a new frame appears in a
[`MediaStreamTrack`](https://developer.mozilla.org/en-US/docs/Web/API/MediaStreamTrack).
This is useful when you need to capture a video stream from a camera or the
screen.

```js
let frames_from_stream = [];
let stream = await navigator.mediaDevices.getUserMedia({ … });
let vtr = new VideoTrackReader(stream.getVideoTracks()[0]);
vtr.start((frame) => {
  frames_from_stream.push(frame);
});
```

<figure class="w-figure">

>   <img src="./consuming-frames-from-tracks.png" alt="The path from ImageBitmap to the network or to storage.">

Reviewer 3: ```suggestion
  <img src="./consuming-frames-from-tracks.png" alt="The path from MediaStreamTrack to the network or to storage.">
```

{% include 'content/origin-trial-register.njk' %}

## Feedback {: #feedback }

> The Chrome team wants to hear about your experiences with the Idle Detection API.

Reviewer: WebCodecs

> The Chrome team wants to hear about your experiences with the Idle Detection API.

Reviewer: ```suggestion
The Chrome team wants to hear about your experiences with the WebCodecs API.
```

### Tell us about the API design

Is there something about the API that doesn't work like you expected? Or are
there missing methods or properties that you need to implement your idea? Have a
question or comment on the security model? File a spec issue on the
corresponding [GitHub repo](https://github.com/WICG/web-codecs/issues), or add
your thoughts to an existing issue.

### Report a problem with the implementation

Did you find a bug with Chrome's implementation? Or is the implementation
different from the spec? File a bug at [new.crbug.com](https://new.crbug.com).
Be sure to include as much detail as you can, simple instructions for
reproducing, and enter `Blink>Media>WebCodecs` in the **Components** box.
[Glitch](https://glitch.com/) works great for sharing quick and easy repros.

### Show support for the API

> Are you planning to use the Idle Detection API? Your public support helps the

Reviewer: WebCodecs

> Are you planning to use the Idle Detection API? Your public support helps the

Reviewer: WebCodecs

Reviewer 3: ```suggestion
Are you planning to use the WebCodecs API? Your public support helps the
```

Chrome team to prioritize features and shows other browser vendors how critical
it is to support them.

Share how you plan to use it on the [WICG Discourse thread][wicg-discourse] Send

> a Tweet to [@ChromiumDev][cr-dev-twitter] with the `#idledetection` hashtag and

Reviewer: ```suggestion
a Tweet to [@ChromiumDev][cr-dev-twitter] with the `#webcodecs` hashtag and
```
